File: p2-isomorphic-strings.cpp
//Time Complexity : O(n) where n is the length of input string
//Space Complexity : O(3n)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : Yes, I had to use 3rd map to save the pair of chars from 's' and 't'.

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        map<char, int> s_map, t_map;
        map<char, char> match;
        map<char, int>::iterator itr1, itr2;
        
        for(int i = 0; i < s.size(); i++){
            itr1 = s_map.find(s[i]);
            itr2 = t_map.find(t[i]);
            
            if(itr1 == s_map.end() && itr2 == t_map.end()){
                s_map.insert(make_pair(s[i], 1));
                t_map.insert(make_pair(t[i], 1));
                match.insert(make_pair(s[i], t[i]));
            }
            else if(itr1 != s_map.end() && itr2 != t_map.end()){
                if(match[s[i]] != t[i])
                    return false;
            }
            else
                return false;
        }
        return true;
    }
};

File: p2_isomorphicStrings.py
# Return if length doesn't match
# Create 2 hashmaps to check one char map of one string with another
# Check if char already exist and match with char from t; return false if it doesn't match else add if the char s doesn't exist in the hashmap
# do the same for mapping from string t to s

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        
        if len(s) != len(t):
            return False

        charMap_st,charMap_ts = {},{}

        for char_s, char_t in zip(s, t):
            
            # Check mapping from string s to t
            if char_s in charMap_st:
                if charMap_st[char_s] != char_t:
                    return False
            else:
                charMap_st[char_s] = char_t

            # Check mapping from string t to s
            if char_t in charMap_ts:
                if charMap_ts[char_t] != char_s:
                    return False
            else:
                charMap_ts[char_t] = char_s

        return True

# ------------------------------------------------------------------------------------------------------------------------------
# Time complexity: O(N) where N is the length of s/t